imbi/endpoints/project_ssm_configuration.py

```diff
@@ -0,0 +1,211 @@
+                    raise web.HTTPError(
+                        401,
+                        reason=
+                        'Not authorized to access SSM in this namespace & environment'
+                    )
+                else:
+                    self.logger.error('Unexpected AWS credential failure: %s',
+                                      data)
+                    raise web.HTTPError(
+                        500,
+                        reason='Unexpected failure fetching AWS credentials')
+            params = aws.get_parameters_by_path(ssm_path_prefix,
+                                                data['access_key_id'],
+                                                data['secret_access_key'],
+                                                data['session_token'])
+            for param in params:
+                if param['Name'] in params_by_path:
+                    params_by_path[
+                        param['Name']]['values'][environment] = param['Value']
+                else:
+                    params_by_path[param['Name']] = {
+                        'type': param['Type'],
+                        'values': {
+                            environment: param['Value']
+                        }
+                    }
+        if not role_arn_exists:
+            raise web.HTTPError(
+                401,
+                reason='No role ARNs exist for this namespace & environment')
+
+        output = [{
+            'name': name,
+            'type': data['type'],
+            'values': [{
+                'environment': environment,
+                'value': value
+            } for environment, value in data['values'].items()]
+        } for name, data in params_by_path.items()]
+        self.send_response(output)
+
+    async def get_aws_credentials(
+            self, role_arn, imbi_user: user.User, id_token,
+            refresh_token) -> tuple[bool, typing.Union[dict, str]]:
+        """Return whether the fetch succeeded, and the data or fail message"""
+        key = f'{imbi_user.username}:{role_arn}'
+        cached_creds = await self._redis.hgetall(key)
+        if cached_creds:
+            return True, {
+                key.decode('utf-8'): value.decode('utf-8')
+                for key, value in cached_creds.items()
+            }
+        try:
+            response = aws.get_credentials(role_arn, imbi_user.username,
+                                           id_token)
+        except botocore.exceptions.ClientError as error:
+            if error.response['Error']['Code'] == 'ExpiredTokenException':
+                id_token = await self.refresh_oauth2_tokens(
+                    imbi_user.username, imbi_user.external_id, refresh_token)
+                return await self.get_aws_credentials(role_arn, imbi_user,
+                                                      id_token, refresh_token)
+            else:
+                return False, error.response['Error']['Code']
```

Comment:
Convert this from recursion to a loop. We might want to set a limit on the number of attempts as well. I'm always concerned about infinite recursion / infinite loops when it comes to external API usage since changes in error responses can cause a lot of grief. This is one of the places where it pays to be a little more defensive.

Reply:
Added a flag on the function. It's still technically recursive, but only retries once as the flag gets flipped.